cha/implementaion/multidevice/our_communication.tex

```diff
@@ -0,0 +1,197 @@
+    \resizebox{!}{0.90\textwidth}{
+        \begin{sequencediagram}
+            \newthread{s}{Server}
+            \newinst{sn}{Server Network}
+            \newinst[2]{wd}{WiFi Direct}
+            \newinst[2]{cn}{Client Network}
+            \newthread{c}{Client}
+
+            \begin{messcall}{s}{Start}{sn}{}
+                \begin{messcall}{sn}{\shortstack{Announce \& \\\\Create Group}}{wd}{}
+                    \begin{messcall}{c}{Start}{cn}{}
+                        \begin{call}{cn}{Discover}{wd}{}
+                        \end{call}
+                        \begin{messcall}{cn}{Services}{c}{}
+                        \end{messcall}
+                        \begin{call}{c}{Connect}{cn}{Network Info}
+                            \begin{call}{cn}{Join Group}{wd}{Network Info}
+                                \begin{call}{wd}{Accept?}{sn}{Network Info}
+                                    \begin{call}{sn}{Accept?}{s}{ACK}
+                                    \end{call}
+                                \end{call}
+                            \end{call}
+                        \end{call}
+                        \begin{messcall}{cn}{\shortstack{Establish\\\\Socket Connection}}{wd}{}
+                            \begin{messcall}{wd}{\shortstack{Accept\\\\Socket Connection}}{sn}{}
+                            \end{messcall}
+                        \end{messcall}
+                    \end{messcall}
+                \end{messcall}
+            \end{messcall}
+        \end{sequencediagram}
+    }
+    \caption{Sequence diagram, depicting how our the server announces its service and clients connects.}\label{fig:seq_server_client}
+\end{figure}
+
+
+\subsection{Establishing a Persistent Socket}
+
+After the client(s) have discovered the server, they establish a persistent connection.
+They are at this point part of the same WiFi Direct network, and the client has just received the IP of the group owner.
+
+When a device first starts its server mode, it launches a separate thread which opens a \code{ServerSocket} to accept incoming socket connections.
+After accepting a connection it creates a MessageConnection, which is our abstraction, which is run on a separate thread for each connection from the server to the client, and vice versa.
+The thread then continuously tries to read on the socket, to accept messages sent to it.
+
+\subsubsection*{Creating our Protobufs}
+As previously mentioned we decided to use protobufs to serialize our messages.
+
+We decided to use ``wire''\footnote{\url{https://github.com/square/wire}} made by ``square'' to generate Java classes corresponding to our protobufs.
+It is a mature library, and easily integrated into our app, it supports both proto2 and proto3, and generates builders for each of the classes it generates, making it easy to use.
+The protobufs we use can be separated into three groups: Meta-data, commands and data. \tnote{Tids ting går her under meta-data, er det fair?}
+\begin{description}
```

Comment:
It is mentioned that we send timestamps in regards to when to play something, but it is never mentioned how a client knows what offset they have to apply, nor where we explain this